core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/EventContextServices.java

```diff
@@ -0,0 +1,414 @@
+    computeTimeLapseStartAndEndCache(scenario);
+
+    // compute Execution/StateFragment and InteractionFragment structure
+    Stream<InteractionFragment> ends = scenario.getOwnedInteractionFragments().stream()
+        .filter(frag -> frag instanceof AbstractEnd || frag instanceof InteractionState);
+    ends.forEachOrdered(end -> {
+      InstanceRole covered = getCoveredInstanceRole(end);
+      if (covered != null && covered.equals(instanceRole)) {
+        Optional<TimeLapse> timeLapse = getTimeLapseFromCache(end);
+
+        // Execution End case
+        if (end instanceof ExecutionEnd) {
+          visit(ancestors, result, (ExecutionEnd) end, timeLapse);
+        }
+
+        // Interaction State case
+        if (end instanceof InteractionState) {
+          visit(ancestors, result, (InteractionState) end, timeLapse);
+        }
+
+        // Message End case
+        if (end instanceof MessageEnd) {
+          visit(ancestors, result, (MessageEnd) end, timeLapse);
+        }
+
+      }
+    });
+    return result;
+  }
+
+  /**
+   * Compute EventStructure for MessageEnd.
+   * 
+   * @param ancestors
+   *          Deque<EObject>
+   * @param result
+   *          List<EventContext>
+   * @param end
+   *          MessageEnd
+   */
+  private static void visit(Deque<CapellaElement> ancestors, List<EventContext> result, MessageEnd end,
+      Optional<TimeLapse> timeLapse) {
+
+    // Execution start / SyncCall and ASyncCall
+    if (timeLapse.isPresent() && timeLapse.get() instanceof Execution
+        && ((Execution) timeLapse.get()).getStart() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), true, ancestors.size() + 1));
+      ancestors.push(timeLapse.get());
+    }
+
+    // Message - EventContext to directly know which execution/instanceRole DNode must be source/target of the
+    // message's DEdge
+    // Handle all cases :
+    // - sending end of the main branch (no execution found in endToEventCache)
+    // - receiving end of the main branch (start of execution found in endToEvent cache and pushed on the ancestor
+    // stack)
+    // - sending end of the return branch (end of execution found in endToEvent cache, not yet removed from the
+    // ancestor cache)
+    // - receiving end of the return branch (no execution found in endToEvent cache).
+    SequenceMessage message = end.getMessage();
+    if (message != null) {
+      result.add(new EventContext(ancestors.peek(), message, end.equals(message.getSendingEnd()), ancestors.size()));
+    }
+
+    // Execution with return branch end / SyncCall
+    if (timeLapse.isPresent() && timeLapse.get() instanceof Execution
+        && ((Execution) timeLapse.get()).getFinish() == end) {
+      ancestors.pop();
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), false, ancestors.size() + 1));
+    }
+
+    // @formatter:off
+        //  The Following diagram would result in
```

Comment:
Thanks for the explication, this is nice :+1:

Reply:
It was hard ten years after the main developments on Sequence to bring all memories together, it will be easier with this.